lib/3scale_toolbox/commands/copy_command/copy_service.rb

```diff
@@ -23,183 +38,263 @@ def self.command
           end
+        end
 
+        def create_mapping_rules
           puts "missing #{missing_mapping_rules.size} mapping rules"
-
           missing_mapping_rules.each do |mapping_rule|
             mapping_rule.delete('links')
             mapping_rule['metric_id'] = metrics_mapping.fetch(mapping_rule.delete('metric_id'))
-            client.create_mapping_rule(service_copy_id, mapping_rule)
+            copy_remote.create_mapping_rule(copy_service_id, mapping_rule)
           end
           puts "created #{missing_mapping_rules.size} mapping rules"
+        end
 
-          puts "extra #{unique_mapping_rules_copy.size} mapping rules"
-          puts unique_mapping_rules_copy.each{|rule| rule.delete('links') }
+        def create_plan_mapping
+          @application_plan_mapping = copy_remote.list_service_application_plans(copy_service_id).map do |plan_copy|
+            plan = source_plans.find do |p|
+              p.fetch('system_name') == plan_copy.fetch('system_name')
+            end
+            [plan['id'], plan_copy['id']]
+          end
+        end
 
-          application_plan_mapping.each do |original_id, copy_id|
-            limits = source_client.list_application_plan_limits(original_id)
-            limits_copy = client.list_application_plan_limits(copy_id)
+        def missing_limits(limits, limits_copy)
+          limits.reject do |limit|
+            limits_copy.find do |limit_copy|
+              limit.fetch('period') == limit_copy.fetch('period')
+            end
+          end
+        end
 
-            missing_limits = limits.reject { |limit| limits_copy.find{|limit_copy| limit.fetch('period') == limit_copy.fetch('period') } }
+        def create_limits
+          application_plan_mapping.each do |original_id, copy_id|
+            limits = source_remote.list_application_plan_limits(original_id)
+            limits_copy = copy_remote.list_application_plan_limits(copy_id)
 
-            missing_limits.each do |limit|
+            m_l = missing_limits(limits, limits_copy)
+            m_l.each do |limit|
               limit.delete('links')
-              client.create_application_plan_limit(copy_id, metrics_mapping.fetch(limit.fetch('metric_id')), limit)
+              copy_remote.create_application_plan_limit(
+                copy_id,
+                metrics_mapping.fetch(limit.fetch('metric_id')),
+                limit
+              )
             end
-            puts "copied application plan #{copy_id} is missing #{missing_limits.size} from the original plan #{original_id}"
+            puts "copied application plan #{copy_id} is missing #{m_l.size} from the original " \
+                 "plan #{original_id}"
           end
         end
+
+        def process
+          create_remotes
+          fetch_source_service
+          create_service
+          fetch_source_proxy_settings
+          create_proxy_settings
+          fetch_source_metrics_methods
+          fetch_copy_metrics_methods
```

Comment:
I think this is insane.

There is an undescribed order between these 18 method calls. Also, because all of them mutate the object's state somehow, it is really mind-boggling to see when is what available. Some of the methods prepare data for later commands, some actually do external actions.

IMO this can be refactored to several smaller classes that each do its part and make the state more immutable by defining the source/copy objects and loading all necessary information upfront.